### ISSUE TITLE
The app references non-public symbols: _getcontext, _makecontext, _setcontext.

### ISSUE THREAD

**Issue by gunerguney** [kivy/kivy-ios#492]
Hi;

In my application I developed, I communicate with api server using openssl.
 
I have successfully compiled my application for iOS devices and tested it on both the simulator and the real device, it works successfully.

However, when i want to deploy the application to app store connect i get the following error and my application is rejected.

> ITMS-90338: Non-public API usage - The app references non-public symbols in ...: **_getcontext, _makecontext, _setcontext.** If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. 

I am reading that it is from **openssl** when I search for the cause of this error. It is mentioned that no-async can be added to the configuration parameters of openssl while building it. Is there anyone who can help me on how to do this?

Regards

**Versions**

* Python : 3.8
* MacOS version : 10.15.1
* XCode Version : 11.3.1
* Cython version : 0.29.17

**Comment by Zen-CODE** [comment]
Are you sure these references come from openssl? Is that all the information it provides? It would be really help if you could provide exact details of where those references come from?

**Comment by gunerguney** [comment]
Thank you @Zen-CODE 

Yes, a developer had already reported the problem in the issues of official openssl GitHub and some people recommended to build without "async".

- https://github.com/openssl/openssl/issues/2545

By the way, this problem has already been experienced in other environments that use openssl about the subject. It was stated in the forums that the problem was solved when it was compiled without async.

 - https://stackoverflow.com/questions/58214571/ios-here-sdk-nmakit-private-api-usage 
 - http://gstreamer-devel.966125.n4.nabble.com/iOS-AppStore-non-public-API-usage-td4687120.html
 - https://github.com/openweave/openweave-core/commit/c9e3bc4a032003fe0a5dfb6c49687224c5f334d2

There are only recommends in these forums:

 - https://forums.developer.apple.com/thread/127678
 - https://forums.developer.apple.com/thread/103286


I found the exact locations of these references when I searched in Kivy-ios folder. 

/kivy-ios/build/openssl/arm64/openssl-1.1.1/crypto/async/arch/async_posix.h
/kivy-ios/build/openssl/arm64/openssl-1.1.1/crypto/async/arch/async_posix.c

and

/kivy-ios/build/openssl/x86_64/openssl-1.1.1/crypto/async/arch/async_posix.h
/kivy-ios/build/openssl/x86_64/openssl-1.1.1/crypto/async/arch/async_posix.c

I wonder if this information is sufficient, can we produce a solution based on these recommendations?

[arch.zip](https://github.com/kivy/kivy-ios/files/4656652/arch.zip)

**Comment by Zen-CODE** [comment]
@gunerguney That's really helpful information. We are looking into it...:-)

**Comment by misl6** [comment]
Hi @gunerguney I just tested it and this issue isn't affecting me.
I suppose that the error `ITMS-90338` it's displaying during upload to App Store.

Which recipes are you packaging into your app?

**Comment by gunerguney** [comment]
Hi @misl6,

I build the kivy ios using the "openssl and kivy" recipes using the command below.
 - toolchain build python3 openssl kivy

I am not using another recipe and the application is running successfully.

I can build and test on a device my app successfully, there is no problem with it. 

When I try to install the application to app store connect, 
 - I archive the Xcode application 
 - validate the app then
 - distribute app to app store, so far there is no problem.

After 4 or 5 minutes, I receive a mail saying that my application was rejected due to the "ITMS-90338: Non-public API usage" from the app store.

**Comment by misl6** [comment]
Can you please try to post the output of `nm -u lib/libssl.a | grep context` ? (the `lib` folder it's Into your kivy-ios's `dist` folder)

**Comment by gunerguney** [comment]
When I run the command "nm -u lib/libssl.a | grep context", I got the following outputs:

no symbols
_tls_validate_all_contexts
no symbols

**Comment by misl6** [comment]
I can suggest 3 things:
- Set deployment target to > `11.0`
- Upgrade to XCode `11.4.x`
- Post full build logs.

**Comment by gunerguney** [comment]
Hi @Zen-CODE ,

Is there any improvement regarding the compilation of openssl without "async" flag?

**Comment by Zen-CODE** [comment]
@gunerguney  Have you tried the suggestions @misl6  suggested? And the full logs if so?

**Comment by gunerguney** [comment]
Yes @Zen-CODE i have tried the @misl6 's suggestions. I had to upgrade my macos version to install xcode 11.4.1 so;

Versions

Python : 3.8
MacOS version : 10.15.5
XCode Version : 11.4.1
Cython version : 0.29.17

Deployment target: ios 13

and I added the full log file.
[Build xpanel_2020-06-28T00-46-02.txt](https://github.com/kivy/kivy-ios/files/4841300/Build.xpanel_2020-06-28T00-46-02.txt)

Thank you

**Comment by Zen-CODE** [comment]
Great, so "build succeeded" from the logs? Can  we close this ticket?

**Comment by gunerguney** [comment]
Hi @Zen-CODE ;

No problem did not solve. I already have no problems with the compilation part. 
When I compile the application and upload it to the app store connect with Xcode, Apple says this error and reject my project.

**Comment by jacob-strauss** [comment]
@gunerguney did you ever figure out how to solve this issue?

I am experiencing a similar problem. I have an identical build to yours (with openssl), the app runs on the simulators and on an actual device, and passes the validation check when archiving the Xcode project. I even have similar specs:

Python : 3.7.7
MacOS version : 10.15.7
XCode Version : 12.0.1
Cython version : 0.29.17

My error message reads:

> ITMS-90338: Non-public API usage - The app references non-public symbols in cccgraphs4: _ffi_call, _ffi_closure_alloc, _ffi_closure_free, _ffi_prep_cif, _ffi_prep_closure_loc, _ffi_type_double, _ffi_type_float, _ffi_type_pointer, _ffi_type_sint16, _ffi_type_sint32, _ffi_type_sint64, _ffi_type_sint8, _ffi_type_uint16, _ffi_type_uint32, _ffi_type_uint64, _ffi_type_uint8, _ffi_type_void. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. 

No idea how to go about solving this. Did you manage?

**Comment by gunerguney** [comment]
No @jacob-strauss  the problem has not been solved yet. I'm still waiting for kivy-ios developers to solve the problem.

**Comment by misl6** [comment]
@jda5 and @gunerguney can you run `toolchain status` and send back a report?

**Comment by jda5** [comment]
@misl6 

I've just done a totally new build. Deleted my venv, reinstalled the toolchain, re-created the Xcode project, configured the settings in the project, archived the project, and got exactly the same email from app:

> ITMS-90338: Non-public API usage - The app references non-public symbols in ccc-graphs-6: _ffi_call, _ffi_closure_alloc, _ffi_closure_free, _ffi_prep_cif, _ffi_prep_closure_loc, _ffi_type_double, _ffi_type_float, _ffi_type_pointer, _ffi_type_sint16, _ffi_type_sint32, _ffi_type_sint64, _ffi_type_sint8, _ffi_type_uint16, _ffi_type_uint32, _ffi_type_uint64, _ffi_type_uint8, _ffi_type_void.

Running `toolchain status` returns

```
audiostream  - Not built
click        - Not built
curly        - Not built
cymunk       - Not built
distribute   - Not built
ffmpeg       - Not built
ffpyplayer   - Not built
flask        - Not built
freetype     - Build OK (built at 2020-12-08 15:30:08.274553)
host_setuptools - Not built
host_setuptools3 - Build OK (built at 2020-12-08 15:42:42.118474)
hostlibffi   - Build OK (built at 2020-12-08 15:30:32.801639)
hostopenssl  - Build OK (built at 2020-12-08 15:31:56.202652)
hostpython3  - Build OK (built at 2020-12-08 15:41:00.044204)
ios          - Build OK (built at 2020-12-08 15:49:29.453920)
itsdangerous - Not built
jinja2       - Not built
kivent_core  - Not built
kivy         - Build OK (built at 2020-12-08 15:55:46.696857)
libcurl      - Not built
libffi       - Build OK (built at 2020-12-08 15:33:57.433529)
libjpeg      - Not built
libpng       - Not built
libzbar      - Not built
markupsafe   - Not built
netifaces    - Not built
numpy        - Not built
openssl      - Build OK (built at 2020-12-08 15:36:33.319987)
photolibrary - Not built
pil          - Not built
pillow       - Not built
pkgresources - Not built
plyer        - Not built
pycrypto     - Not built
pykka        - Not built
pyobjus      - Build OK (built at 2020-12-08 15:50:40.436716)
python3      - Build OK (built at 2020-12-08 15:49:09.344737)
pyyaml       - Not built
sdl2         - Build OK (built at 2020-12-08 15:37:40.964816)
sdl2_image   - Build OK (built at 2020-12-08 15:41:38.455632)
sdl2_mixer   - Build OK (built at 2020-12-08 15:42:28.590812)
sdl2_ttf     - Build OK (built at 2020-12-08 15:42:39.149817)
werkzeug     - Not built
zbarlight    - Not built
```

I was maybe thinking of cleaning libffi and hostlibffi and then creating an Xcode project (or updating the old one). Would that work?

Thanks

**Comment by nikosgrikos** [comment]
This looks similar to https://github.com/kivy/kivy-ios/issues/572

**Comment by misl6** [comment]
@gunerguney I'm don't close this issue cause it seems a little bit different from #572 , can you please try to build on top of `kivy-ios==1.2.1` and check if the fix also worked for you?

**Comment by gunerguney** [comment]
I am far away from my MacBook due to covid-19, I did not have remote access. I will try it at the first opportunity and share my results. Thanks

**Comment by gunerguney** [comment]
Finally we managed to recompile the app, uploaded it to app store connect and yesterday review completed successfully. Our application is now live. I would like to state that our problem is solved. We made only one change throughout the application, we were using a self-signed certificate before, and it had to be in the project's folders. Now we have made this a globally signed certificate, I don't know if it had an effect on solving the problem.

Thank you for your help @Zen-CODE @misl6
